ccmlib/cluster.py

```diff
@@ -106,36 +106,78 @@ def set_install_dir(self, install_dir=None, version=None, verbose=False):
 
     def actively_watch_logs_for_error(self, on_error_call, interval=1):
         """
-        Begins a timed daemon thread that scans logs for errors every interval seconds.
+        Begins a thread that scans logs for errors every interval seconds.
 
-        If an error is seen and the callback is executed, it will be called with an
-        OrderedDictionary mapping node name to a list of error lines.
+        If an error is seen, the callback is executed with an OrderedDictionary
+        mapping node name to a list of error lines.
+
+        Returns the thread itself, which should be .join()'ed to wrap up execution,
+        otherwise will run until the main thread exits.
         """
-        log_positions = defaultdict(int)
+        class LogWatchingThread(threading.Thread):
+            """
+            This class is embedded here for now, because it is used only from
+            within Cluster, and depends on cluster.nodelist().
+            """
+            def __init__(self, cluster):
+                super(LogWatchingThread, self).__init__()
+                self.cluster = cluster
+                self.daemon = True  # set so that thread will exit when main thread exits
+                self.req_stop_event = threading.Event()
+                self.done_event = threading.Event()
+                self.log_positions = defaultdict(int)
+
+            def scan(self):
+                errordata = OrderedDict()
 
-        def grep_and_mark_all_logs():
-            errordata = OrderedDict()
+                try:
+                    for node in self.cluster.nodelist():
+                        errors = node.grep_log_for_errors_from(seek_start=self.log_positions[node.name])
+                        self.log_positions[node.name] = node.mark_log()
+                        if errors:
+                            errordata[node.name] = errors
```

Comment:
no, the dict used is short-lived, and just represents errors for that pass. each node is checked once, and keyed by its name, so unless nodes somehow collide names I think this is safe.